index.html

```
@@ -839,33 +841,33 @@
               <var>acceptPromise</var> with that error and terminate this
               algorithm.
               </li>
-              <li>Determine which <a>payment handlers</a> support
-              <var>identifier</var>. For each resulting payment handler, if
-              payment method specific capabilities supplied by the payment
-              handler match those provided by <var>data</var>, the payment
-              handler matches.
+              <li>Add each <a>payment handler</a> that supports
```

This is much clearer wording, nice!

BUT, it does ignore the case where the browser is able to fail fast because the handler allows it to. 

The argument that the payment sheet is still shown is correct BUT that is a side-effect of how the `basic-card` payment handler is implemented in those browsers.
i.e. It claims that it supports all networks and card types so it always matches. 

Those implementations have decided that "support" does not mean "has a card loaded" but rather "is able to load a card if needed". If we want to be specific about what "support" means then I agree that should be defined in the payment method spec for that payment method. But, we still need a hook here to tell implementors that this is where they do that check.

A future payment method or a different implementation of a `basic-card` handler may have a much more definitive response when determining if the handler supports the request based on what's in the data.

(And I agree that the output of this should be consistent between `show()` and `canMakePayment()`).

Could this read:

```html
<li>For each <a>payment handler</a> that supports <var>identifier</var>:
    <ol>
        <li>If <var>data</var> indicates the <a>payment handler</a> 
        will not be able to handle the payment, continue.</li>
        <li>Else, add <a>payment handler</a> into <var>handlers</var>.</li>
    </ol>
</li>
```

---

I agree with @adrianba's formulation from some time ago:

 1) Build a set of handlers based on PMI matching
 2) Then potentially remove that set based on some other factors

I can think of at least three factors a user agent might taken into account
to remove a payment handler:

 * Payment method specific capability matching
 * Payment method manifest statements
 * Security considerations (e.g., known malicious code or origin)
 
Marcos' statement is broad enough to accommodate these factors:
   "Add each <a>payment handler</a> that supports <var>identifier</var> into <var>handlers</var>."

But I would like to be more explicit about the factors that inform the user agent's decision.
For example:

<pre>
  <li>For each <a>payment handler</a> that supports <var>identifier</var>:
    <ol>
        <li>If the <a>payment handler</a> is not able or authorized to handle
              the payment request, continue. Note: Payment method-specific
             <var>data</var> and information provided through other mechanisms
             inform the user agent's determination.</li>
        <li>Else, add <a>payment handler</a> into <var>handlers</var>.</li>
    </ol>
  </li>
</pre>

---

It would be disingenuous the mention `.data` here, because the steps before this one already rejected bad data explicitly with an error. Let's please avoid making statements about how things work without solid proof that they work that way. 

Hypothetical "well, another user agent might maybe oneday kinda sorta do it differently" strawpersons are extremely unhelpful. Let's please stick to documenting reality.

---

> BUT, it does ignore the case where the browser is able to fail fast because the handler allows it to.

There is no browser today that fails fast. If a browser in the future does decide to fail fast, we should amend the spec then.  

Again, I feel very strongly that we should only document how things work today for the purpose of CR: We can spin up a "V2" very easily in a few weeks, the day _AFTER_ CR, if we want to add any extended behavior. If we try to capture what future browser engines will do now, we will likely get it wrong - so let's do it right when the time comes (in a "V2" please!). 

> The argument that the payment sheet is still shown is correct BUT that is a side-effect of how the basic-card payment handler is implemented in those browsers. i.e. It claims that it supports all networks and card types so it always matches.

Actually, we are both wrong (I accidentally misled you, sorry!): When Chrome doesn't know the network, it restricts to the networks it knows about (Visa, MasterCard, etc.) - while allowing any type (I was surprised by this too! 😱). 

So, if a merchant requests "this-does-not-exists" and/or "voyager", they will likely get back a "visa" card (or something else Chrome supports and the end-user has on hand). If the user doesn't have any other card on hand, they have to abort the payment request. 

I've specified the above here:
https://github.com/w3c/payment-method-basic-card/pull/39/

However, my assertion above still holds. The `.data` doesn't play a role here in today's implementations. It may in the future, but again, please please please 🎶PLLeeaAAAASssseeEE🎶, let's not add aspirational things right before CR. We should be looking at reducing the spec right now, not growing its scope.

---

We discussed this on today's Editor's call. @zkoch gave me a "reality check": when Chrome does work in concert with third-party payment methods (native apps in this case), they do "fail fast" at this point. Given that what I sought was confirmation that another use agent was in fact doing doing the filtering, then I concur and will add the text that @ianbjacobs suggested.

---

Many thanks to the editors for working this out!

Ian